# Introduction to Scikit-Learn (sklearn)

This notebook demonstrates some of the most useful functions of the beautiful Scikit-Lern library

What we're going to cover: 

0. An end-to-end Scikit-Learn workflow
1. Getting the data ready 
2. Choose the right estimator/algorithm for our problem
3. Fit the model/algorithm and use it to make predictions on our data
4. Evaluating a model
5. Improve a model 
6. Save and load a trained model
7. Putting it all together!

## 0. An end-to-end Scikit-Learn workflow 

In [30]:
import numpy as np

In [31]:
# 1. Get the data ready 
import pandas as pd 
heart_disease = pd.read_csv("data/heart-disease.csv")
heart_disease

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0     63    1   3       145   233    1        0      150      0      2.3   
1     37    1   2       130   250    0        1      187      0      3.5   
2     41    0   1       130   204    0        0      172      0      1.4   
3     56    1   1       120   236    0        1      178      0      0.8   
4     57    0   0       120   354    0        1      163      1      0.6   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
298   57    0   0       140   241    0        1      123      1      0.2   
299   45    1   3       110   264    0        1      132      0      1.2   
300   68    1   0       144   193    1        1      141      0      3.4   
301   57    1   0       130   131    0        1      115      1      1.2   
302   57    0   1       130   236    0        0      174      0      0.0   

     slope  ca  thal  target  
0        0   0     1       1  
1        0   0     2       1  
2        2   0     2       1  
3        2   0     2       1  
4        2   0     2       1  
..     ...  ..   ...     ...  
298      1   0     3       0  
299      1   0     3       0  
300      1   2     3       0  
301      1   1     3       0  
302      1   1     2       0  

[303 rows x 14 columns]

In [32]:
# Create X (feature matrix)
X = heart_disease.drop("target", axis=1)

# Create y (labels)
y = heart_disease["target"]

In [33]:
# 2. Choose the right model and hyperparametes (used to tune the data)
# our problem is classification 
from sklearn.ensemble import RandomForestClassifier # Classification ML Model
clf = RandomForestClassifier(n_estimators=100)

# We'll keep the default hyperparameters 
clf.get_params()

{'bootstrap': True,
 'ccp_alpha': 0.0,
 'class_weight': None,
 'criterion': 'gini',
 'max_depth': None,
 'max_features': 'auto',
 'max_leaf_nodes': None,
 'max_samples': None,
 'min_impurity_decrease': 0.0,
 'min_impurity_split': None,
 'min_samples_leaf': 1,
 'min_samples_split': 2,
 'min_weight_fraction_leaf': 0.0,
 'n_estimators': 100,
 'n_jobs': None,
 'oob_score': False,
 'random_state': None,
 'verbose': 0,
 'warm_start': False}

In [34]:
# 3. Fit the model to the [training] data
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

In [35]:
clf.fit(X_train, y_train);

note: If the shape of the data is not same shape, it will throw error

In [36]:
# make a prediction
y_label = clf.predict(np.array([0, 2, 3, 4]))

ValueError: Expected 2D array, got 1D array instead:
array=[0. 2. 3. 4.].
Reshape your data either using array.reshape(-1, 1) if your data has a single feature or array.reshape(1, -1) if it contains a single sample.

In [37]:
y_preds = clf.predict(X_test)
y_preds

array([0, 1, 1, 0, 1, 1, 1, 0, 0, 1, 1, 0, 1, 0, 1, 1, 1, 0, 0, 0, 0, 0,
       1, 1, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 0, 1, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0])

In [38]:
y_test

179    0
228    0
111    1
246    0
60     1
      ..
249    0
104    1
300    0
193    0
184    0
Name: target, Length: 61, dtype: int64

In [39]:
# 4. Evaluate the model on the training data and the test data
clf.score(X_train, y_train)

1.0

In [40]:
clf.score(X_test, y_test)

0.8524590163934426

In [41]:
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score

print(classification_report(y_test, y_preds))

              precision    recall  f1-score   support

           0       0.86      0.83      0.84        29
           1       0.85      0.88      0.86        32

    accuracy                           0.85        61
   macro avg       0.85      0.85      0.85        61
weighted avg       0.85      0.85      0.85        61



In [42]:
confusion_matrix(y_test, y_preds)

array([[24,  5],
       [ 4, 28]])

In [43]:
accuracy_score(y_test, y_preds)

0.8524590163934426

Explain: We are using loop here to test different estimators values into the **training set** to see if it make a different in predicting the **test set**. Thus we are adjusting the model to find the prefect fit 

In [52]:
# 5. Improve a model 
#Try different amount of n_estimators 
np.random.seed(42)
for i in range(10, 100, 10):
    print (f"Tryinng model with {i} estimators...")
    clf = RandomForestClassifier(n_estimators=i).fit(X_train, y_train)
    print(
        f"Model accuracy on test set: {clf.score(X_test, y_test) * 100:.2f}%")
    print(" ")
    

Tryinng model with 10 estimators...
Model accuracy on test set: 85.25%
 
Tryinng model with 20 estimators...
Model accuracy on test set: 80.33%
 
Tryinng model with 30 estimators...
Model accuracy on test set: 83.61%
 
Tryinng model with 40 estimators...
Model accuracy on test set: 80.33%
 
Tryinng model with 50 estimators...
Model accuracy on test set: 86.89%
 
Tryinng model with 60 estimators...
Model accuracy on test set: 83.61%
 
Tryinng model with 70 estimators...
Model accuracy on test set: 83.61%
 
Tryinng model with 80 estimators...
Model accuracy on test set: 83.61%
 
Tryinng model with 90 estimators...
Model accuracy on test set: 81.97%
 


In [54]:
# 6. Save a model and load it 
import pickle 
# wb = write binary
pickle.dump(clf, open("random_forst_model_1.pkl", "wb"))

Note: The saved model will be the last estimated which is 81.97%

In [55]:
loaded_model = pickle.load(open("random_forst_model_1.pkl", "rb"))
loaded_model.score(X_test, y_test)

0.819672131147541